### PR TITLE
Refactor plugin loading

### PR DESCRIPTION
**Pull request by kvosper**
The primary focus of this PR is to decouple the Admin server from the plugin-loading, so that will be able to visit the admin server before the plugins have all been loaded.
That work is not totally finished by this PR (as it is already quite large and we don't want to add even more changes).

I have also taken the opportunity to tidy-up parts of the plugin-loading code that were messy.

It is a prerequisite to allowing Styx to have proper liveness/readiness checks, which needs the admin server to be able to start without waiting for the plugins.

### CODE REVIEW

File: components/proxy/src/main/java/com/hotels/styx/startup/extensions/FailureHandling.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+final class FailureHandling {
+    private static final Logger LOGGER = getLogger(FailureHandling.class);
+
+    static final FailureHandlingStrategy<Pair<String, SpiExtension>, ConfiguredPluginFactory> PLUGIN_FACTORY_LOADING_FAILURE_HANDLING_STRATEGY =
+            new FailureHandlingStrategy.Builder<Pair<String, SpiExtension>, ConfiguredPluginFactory>()
+
+                    .doImmediatelyOnEachFailure((plugin, err) -> logIndividualFailure(new FailedPlugin(plugin), err))
+
+                    .doOnFailuresAfterAllProcessing(failures -> {
+                        throw new PluginStartupException(afterFailuresErrorMessage(transformKeys(failures, FailedPlugin::new)));
```
Comment (dvlato) [inline components/proxy/src/main/java/com/hotels/styx/startup/extensions/FailureHandling.java]: it seems transformKeys()'s second argument is always  "FailedPlugin::new" . Then why do we need it be an argument at all?

File: components/proxy/src/main/java/com/hotels/styx/startup/extensions/FailureHandling.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+        return format("%s plugin%s could not be loaded: failedPlugins=%s; failureCauses=%s",
+                failures.size(),
+                failures.size() == 1 ? "" : "s", // only pluralise plurals
+                failedPlugins,
+                causes
+        );
+    }
+
+    private static class FailedPlugin {
```
Comment (dvlato) [inline components/proxy/src/main/java/com/hotels/styx/startup/extensions/FailureHandling.java]: Nitpicking again, but I wonder if we really need a class for this purpose.

File: components/proxy/src/main/java/com/hotels/styx/startup/extensions/PluginLoadingForStartup.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    /**
+     * Load plugins.
+     *
+     * @param environment environment
+     * @param factories   if set, overrides config, otherwise plugins will be loaded from config
+     * @return plugins
+     */
+    public static List<NamedPlugin> loadPlugins(Environment environment, List<ConfiguredPluginFactory> factories) {
+        if (factories == null) {
```
Comment (dvlato) [inline components/proxy/src/main/java/com/hotels/styx/startup/extensions/PluginLoadingForStartup.java]: Same comment about null arguments.

File: components/proxy/src/main/java/com/hotels/styx/startup/StyxServerComponents.java [hunk 4]
```diff
@@ -174,7 +184,7 @@ public StyxServerComponents build() {
     public interface LoggingSetUp {
         LoggingSetUp DO_NOT_MODIFY = environment -> {
         };
-        LoggingSetUp FROM_CONFIG = environment -> setUpLogging(environment.configuration().logConfigLocation());
+        LoggingSetUp FROM_CONFIG = environment -> setUpLogging(environment.styxConfig().logConfigLocation());
```
[the remark this answers is not in the document]
Comment (kvosper) [reply components/proxy/src/main/java/com/hotels/styx/startup/StyxServerComponents.java]: Both of these methods already existed. This is a change to which method is called. I will try to unify it though, as we probably don't need both.

File: components/proxy/src/main/java/com/hotels/styx/startup/extensions/ConfiguredPluginFactory.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    private final Function<Class<?>, Object> configProvider;
+
+    public ConfiguredPluginFactory(String name, PluginFactory pluginFactory, Function<Class<?>, Object> configProvider) {
+        this.name = requireNonNull(name);
+        this.pluginFactory = requireNonNull(pluginFactory);
+        this.configProvider = configProvider == null ? any -> null : configProvider;
+    }
+
+    public ConfiguredPluginFactory(String name, PluginFactory pluginFactory) {
+        this(name, pluginFactory, null);
```
Comment (dvlato) [inline components/proxy/src/main/java/com/hotels/styx/startup/extensions/ConfiguredPluginFactory.java]: Why don't you add the "any -> null" here so you can enforce that configProvider is not null in above's constructor?  I don't know if it requires changing anything other than a few Tests...

**Pull request merged**